src/configuration_server.erl

```diff
@@ -0,0 +1,160 @@
+handle_call(default_directories, _From, State = #configuration{}) ->
+    {reply, {ok, default_directories()}, State};
+handle_call(_Request, _From, State = #configuration{}) ->
+    {reply, ok, State}.
+
+%% @private
+%% @doc Handling cast messages
+-spec handle_cast(Request :: term(), State :: #configuration{}) ->
+    {noreply, NewState :: #configuration{}} |
+    {noreply, NewState :: #configuration{}, timeout() | hibernate} |
+    {stop, Reason :: term(), NewState :: #configuration{}}.
+handle_cast(_Request, State = #configuration{}) ->
+    {noreply, State}.
+
+%% @private
+%% @doc Handling all non call/cast messages
+-spec handle_info(Info :: timeout() | term(), State :: #configuration{}) ->
+    {noreply, NewState :: #configuration{}} |
+    {noreply, NewState :: #configuration{}, timeout() | hibernate} |
+    {stop, Reason :: term(), NewState :: #configuration{}}.
+handle_info(_Info, State = #configuration{}) ->
+    {noreply, State}.
+
+%% @private
+%% @doc This function is called by a gen_server when it is about to
+%% terminate. It should be the opposite of Module:init/1 and do any
+%% necessary cleaning up. When it returns, the gen_server terminates
+%% with Reason. The return value is ignored.
+-spec terminate(
+    Reason :: (normal | shutdown | {shutdown, term()} | term()),
+    State :: #configuration{}
+) ->
+    term().
+terminate(_Reason, _State = #configuration{}) ->
+    ok.
+
+%% @private
+%% @doc Convert process state when code is changed
+-spec code_change(
+    OldVsn :: term() | {down, term()},
+    State :: #configuration{},
+    Extra :: term()
+) ->
+    {ok, NewState :: #configuration{}} | {error, Reason :: term()}.
+code_change(_OldVsn, State = #configuration{}, _Extra) ->
+    {ok, State}.
+
+%%%===================================================================
+%%% Internal functions
+%%%===================================================================
+%% @doc This function returns the predefined configuration directories.
+-spec default_directories() ->
+    [file:filename_all()].
+default_directories() ->
+    ?CONFIG_PATHS.
+
+%% @doc This function returns the predefined filenames for config files.
+-spec default_filenames() ->
+    [file:filename_all()].
+default_filenames() ->
+    ?FILE_NAMES.
+
+%% @private
+%% @doc This function searches for a configuration file in the default
+%% directories and uses the first found file.
+-spec choose_config_file(
+    Directories :: [file:filename_all()],
+    FileNames :: [file:filename_all()]
+) ->
+    {ok, FileName :: file:filename_all()} | {error, Reason :: term()}.
+choose_config_file(Directories, FileNames) ->
+    CandidateNames =
+        [filename:absname_join(Path, File) || Path <- Directories, File <- FileNames],
+    case [File || File <- CandidateNames, filelib:is_regular(File)] of
+        ConfigFiles when length(ConfigFiles) >= 1 -> {ok, lists:nth(1, ConfigFiles)};
+        _ -> {error, {no_file, "No config file found"}}
+    end.
```

Comment:
Again I'm reminded of https://github.com/inaka/erlang_guidelines/blob/master/src/smaller_functions.erl

Reply:
Either way I personally don't find this very readable. So even if the `case` is something we can't get rid of, I think this is too terse as is.

Reply:
This has been changed to work without `case`.